scripts/solve.js

```diff
@@ -175,3 +187,45 @@ async function submitSolution() {
   submitButton.innerText    = "Submit";
   submitButton.disabled     = false;
 }
+
+/**
+ * Helper function to calculate position in a codemirror instance.
+ * Takes in a character position given by JS-interpreter and
+ * calculates its line and character position in a codemirror instance.
+ * Returns position as a {line: num, ch: num} object as needed by codemirror.
+ **/
+function calculatePosition(charPosition) {
+  let doc = codeMirror.getDoc();
+
+  for (let i = 0; i < doc.lineCount(); ++i) {
+    // Adds + 1 to account for newline character
+    let currentLineLength = doc.getLine(i).length + 1;
+
+    if (currentLineLength >= charPosition) {
+      return {line: i, ch: charPosition};
+    } else {
+      charPosition -= currentLineLength;
+    }
+  }
+}
+
+/**
+ * Highlights code section that the interpreter is currently parsing.
+ **/
+function highlightCode(interpreter) {
+  let start = 0;
+  let end = 0;
+
+  if (interpreter.stateStack.length) {
+    let node    = interpreter.stateStack[interpreter.stateStack.length - 1].node;
+    start       = node.start;
+    end         = node.end;
+  }
```

Comment:
could you explain what this section of the code is doing? lgtm overall!

Reply:
This is for getting what section of the code the interpreter is currently processing.
To be perfectly honest, I'm not perfectly sure what `stateStack` is about as documentation is scarce.

Looking at the way it works while executing the code, my understanding is that `stateStack` is the code [execution stack](https://blog.bitsrc.io/understanding-execution-context-and-execution-stack-in-javascript-1c9ea8642dd0). The `.node.start` and `.node.end` simply tells us what section of the code the execution stack is currently at.

Reply:
good explanation; to add on, a `call stack` is a common concept in programming languages that spans from high-level programming languages to assembly.

a good SO [explanation](https://stackoverflow.com/questions/10057443/explain-the-concept-of-a-stack-frame-in-a-nutshell#:~:text=A%20stack%20frame%20is%20a,and%20space%20for%20local%20variables.)

Reply:
ty both! @Nicholas2750 @chronologos